### DISCUSSION TITLE
Request: Much more dynamic form

### DISCUSSION THREAD

**Discussion by arladmin** [Budibase/budibase#1617] [Ideas]
As forms are the primary route for data entry into the system, i would suggest forms to have more dynamic capabilities. Some of the desired capabilities, which i'm already used to are:
- Dynamic Rules to control visibility of fields (ability to set conditions like: if value of another field is x then show/hide, or if the logged-in User's y field value is z then show/hide, etc.)
- Select the appropriate Data source (within the same table):
e.g. The form is for a table called Actions which includes a Relationship Picker for another table, ToDo.
The ToDo table includes fields like (1. User Relationship - 'Created By', 2. User Relationship - 'Shared with', Table Relationship - 'Team') and the User table already has a field like (User connected with a Team table)

   Now, when configuring a new Form for Actions, i would like to be able to set, for the ToDos field, criteria for the data source:
   - ToDos created by the logged-in User
   - ToDos shared with' the logged-in User
   - ToDos linked with the same Team as the logged-in User  
   - All available ToDos
- Limit (Filter) the options for fields, especially fields like Date Picker, Options Picker and Relationship Picker (e.g. from the previously selected data source, allow only options that meet a certain criteria or show dates within a particular range(based on other fields) etc. )
- Possibility to add a (popup) sub-form to add missing/extra options in Options Picker/Relationship Picker fields (again, the sub-form should be equally dynamic)

Kindly let me know if some of these features are already available.
Thanks.

**Comment by mjashanks** [comment]
We are planning to address some of these issues in the next release: https://github.com/Budibase/budibase/issues/1644

- Control visibility of fields, based on other fields - YES
- Limit the options for fields - potentially
- Select appropriate Datasource - potentially, but more likely for the future. 

For the sub-form - do you mean a form that could add a new record, that would then be available in the options picker?

**Comment by arladmin** [comment]
@mjashanks 
Here's a small **[video](https://www.dropbox.com/s/kkyysw6u3r9wxon/screen-capture.webm?raw=1)**, of the frontend, depicting the functionality that i mentioned above.

As well as, a screenshot, from the app builder, that shows the Field Filter and the Data Source Filter functionalities:
![image](https://user-images.githubusercontent.com/64059154/121034740-e0934080-c7ca-11eb-95bd-7e6072c7f279.png)

And another screenshot displaying the Field visibility functionality:
![image](https://user-images.githubusercontent.com/64059154/121036080-1ab11200-c7cc-11eb-9b27-cbbc77d9d073.png)

**Reply by mjashanks** [reply]
Thanks so much, this is great!

Would love to more about how you are using Budibase - send me an email if you are up for a chat :)

**Reply by arladmin** [reply]
I shall ping you on Discord.

**Reply by KernOil** [reply]
this functionality is a must have for the BB system
